xds/internal/balancer/cdsbalancer/cdsbalancer.go

```diff
@@ -271,6 +265,45 @@ func (b *cdsBalancer) run() {
 	}
 }
 
+// handleErrorFromUpdate handles both the error from parent ClientConn (from
+// resolver) and the error from xds client (from the watcher). fromParent is
+// true if error is from parent ClientConn.
+//
+// If the error is connection error, it's passed down to the child policy.
+// Nothing needs to be done in CDS (e.g. it doesn't go into fallback).
+//
+// If the error is resource-not-found:
+// - If it's from resolver, it means LDS resources were removed. The CDS watch
+// should be canceled.
+// - If it's from xds client, it means CDS resource were removed. The CDS
+// watcher should keep watching.
+//
+// In both cases, the error will be forwarded to EDS balancer. And if error is
+// resource-not-found, the child EDS balancer will stop watching EDS.
+func (b *cdsBalancer) handleErrorFromUpdate(err error, fromParent bool) {
+	// TODO: connection errors will be sent to the eds balancers directly, and
+	// also forwarded by the parent balancers/resolvers. So the eds balancer may
+	// see the same error multiple times. We way want to only forward the error
+	// to eds if it's not a connection error.
+	//
+	// This is not necessary today, because xds client never sends connection
+	// errors.
+
+	if fromParent && xdsclient.ErrType(err) == xdsclient.ErrorTypeResourceNotFound && b.cancelWatch != nil {
```

Comment:
Optional: Might be a bit more readable to have the check for `cancelWatch` inside the top if.

Reply:
Done